notes/advanced-data-types.md

```diff
@@ -0,0 +1,530 @@
+
+A cool thing, starting in C++11, you can also place functions into `struct`s as well, as the following figure shows:
+```
+struct {
+    int myField1;
+    double myField2;
+    char myFunction() {
+        return 'c';
+    }    
+};
+```
+<div style="text-align: center"> Figure 3 - Struct Declaration and Initialization </div>
+
+## Enumerations (`enum`)
+
+An `enum` is an entity that maps literal constants to integers. 
+
+```
+/*
+enum enumName {
+    constant1,
+    constant2,
+    ...
+    constantN
+};
+*/
+enum weekdays {
+    Sunday,
+    Monday,
+    Tuesday,
+    Wednesday,
+    Thursday,
+    Friday,
+    Saturday
+};
+```
+<div style="text-align: center"> Figure 4a - Enum Declaration </div>
+
+Note how each element does not need a type identifier, they are simply literals, and that they are seperated by (`,`), not (`;`). You will also note there are no explicit values assigned to each constant. That is because if that is not done, the first constant is 0, the second is one more than that, and so on. In this example, this would be the same as saying:
+
+```
+/*
+enum enumName {
+    constant1 = value1,
+    constant2 = value2,
+    ...
+    constantN = valueN
+};
+*/
+
+enum weekdays {
+    Sunday = 0,
+    Monday = 1,
+    Tuesday = 2,
+    Wednesday = 3,
+    Thursday = 4,
+    Friday = 5,
+    Saturday = 6
+};
+```
+<div style="text-align: center"> Figure 4b - Enum Declaration </div>
+
+Now, if you want specific values for each constant, you can use this `= integer` notation, as shown below:
+```
+enum painMeter {
+    noPain = 0,
+    somePain = 3,
+    pain = 5,
+    extremePain = 7,
+    CSE351 = 10
+};
+```
+<div style="text-align: center"> Figure 4c - Enum Declaration </div>
+
+You can also do a mix of both
+
+```
+enum HuskySatTwo {
+    LOST,
+    ReactionWheels = 4,
+    Magnetorquers
+    FOUND,
+    Propulsion = 10
+};
+```
+<div style="text-align: center"> Figure 4d - Enum Declaration </div>
+
+- `LOST` has no intitialized value, so its value is 0
+- `ReactionWheels` has an initialized value, so its value is 4
+- `Magnetorquers` has no initialized value, so it is one more than the last, or 5
+- `FOUND` also has no initialized value, so it is 6
+- `Propulsion` does, so it is 10
+
+After an enum is declared, you can then make variables out of it using the following notation, which is similar to the declaration of a struct
+
+```
+// enumName enumVarName = constant;
+
+weekday day = Wednesday;
+painMeter pain;
+pain = extremePain;
+```
+<div style="text-align: center"> Figure 4e - Enum Initialization </div>
+
+## Enumeration Classes (`enum class`)
+
+However, once declared, enums will not allow for any redefinition of any of its constants while you are in the enum's scope. An enum's scope ranges from its declaration to the `}` brackets that contain it (Or the end of the file), and this applies to its fields as well. The below demonstrates a few errors about this:
+
+```
```

Comment:
btw fun thing about Markdown code blocks, you can write the language after the opening triple backtick, for example:
'''C++
// code
'''

and the code will be highlighted appropriately

Reply:
Thank you!